sharktank/sharktank/ops/sharded_impls.py

```diff
@@ -26,9 +26,188 @@
+            kwargs = unflatten_args(kwargs, t_i_kwargs, t_vals[len(t_vals_args) :])
+            res = f(*args, **kwargs)
+            if isinstance(res, ShardedTensor) and len(t_vals) > 0:
+                pinned = (
+                    res.pinned
+                    or (len(t_vals) == 1 and t_vals[0].pinned)
+                    or "_like_" in f.__name__
+                )  # TODO: How to handle reshard_like (and future similar ones properly)
+                res = res.clone(devices=t_vals[0].devices, pinned=pinned)
+            return res
+
+        return func_wrapper
+
+    def wrap_override(signature_dispatcher_override):
+        """
+        Wrap [op].override's result so that the transfer_n_pin(f) becomes the target in _TargetOverride rather than f itself.
+        """
+
+        def override_return_wrapper(*override_args, **override_kwargs):
+            orig_decorator = signature_dispatcher_override(
+                *override_args, **override_kwargs
+            )
+            new_decorator = lambda f: orig_decorator(transfer_n_pin(f))
+            return new_decorator
+
+        return override_return_wrapper
+
+    do_not_wrap = {
+        "all_gather",
+        "all_reduce",
+        "replicate",
+        "index_copy_",
+        "index_put_",
+        "transfer_to_logical_device",
+    }
+
+    from . import signatures
+
+    for func_name in signatures.__all__:
+        func = globals()[func_name]
+        if (func_name not in do_not_wrap) and (hasattr(func, "override")):
+            func.override_orig = func.override
+            func.override = wrap_override(func.override_orig)
+
+
+def unwrap_override_decorator():
+    """
+    Unwraps [op].override to restore the original function.
+    Must be called at the end of this file.
+    """
+    from . import signatures
+
+    for func_name in signatures.__all__:
+        func = globals()[func_name]
+        if hasattr(func, "override_orig"):
+            func.override = func.override_orig
+            del func.override_orig
+
+
+wrap_override_decorator()
+
+
+def transfer_if_needed(*tensors: Tuple[ShardedTensor]) -> List[ShardedTensor]:
+    """
+    If at least 2 tensors are panned in, the shards of all unpinned tensors are transfered to be on the same devices as those of the pinned tensors.
+    """
+
+    def tensor_w_shards_moved(
+        tensor: ShardedTensor, new_devices: Tuple[int]
+    ) -> ShardedTensor:
+        """
+        Create a copy of the passed in tensor, but with shards now placed on the new devices.
+        """
+        new_shards = tuple(
+            (
+                transfer_to_logical_device(shard, new_devices[j])
+                if new_devices[j] != tensor.devices[j]
+                else barrier_on_logical_device(shard, new_devices[j])
+            )
+            for j, shard in enumerate(tensor.shards)
+        )
+        return tensor.clone(ts=new_shards, devices=new_devices)
```

Comment:
@rsuderman 

What, if anything, do I need to do about `transfrom_globals` and DeviceTensorTraits in this function or in tensor.clone?

Reply:
We wouldn't want to set the `DeviceTensorTrait` in these cases. Those should only be manipulated during placement examples, so no worries on clone.